[RFC] vim-patch:7.4.{1641,1643,1652,1654}

jamessan opened the pull request:
```
Problem:    Using unterminated string.
Solution:   Add NUL before calling vim_strsave_shellescape(). (James McCoy)
```

https://github.com/vim/vim/commit/5ca84ce4aa2832041f843e624c222bbc1f4d3e14

Some of the transformations that occur earlier in modify_fname simply adjust `*fnamelen` to indicate to the caller what portion of `*fnamep` to use as the modified filename.  `*fnamep` needs its NUL-termination to reflect the proper substring indicated by `*fnamelen` before calling `vim_strsave_shellescape`.

jamessan commented on test/functional/legacy/105_filename_modifiers_spec.lua:
```diff
@@ -45,6 +45,7 @@ describe('filename modifiers', function()
     execute([=[Put fnamemodify("abc\ndef",       ':S'      )]=])
     execute([=[set shell=tcsh]=])
     execute([=[Put fnamemodify("abc\ndef",       ':S'      )]=])
+    execute([=[Put expand('%:r:S') == shellescape(expand('%:r'))]=])
```
This line appears to be breaking the test harness.  I get this lua traceback that I can't make any sense out of.

```
Error → test/functional/legacy/105_filename_modifiers_spec.lua @ 10
filename modifiers is working
...m/neovim/.deps/usr/share/lua/5.1/nvim/msgpack_stream.lua:65: end of file

stack traceback:
        ...m/neovim/.deps/usr/share/lua/5.1/nvim/msgpack_stream.lua:65: in function 'run'
        ...om/neovim/.deps/usr/share/lua/5.1/nvim/async_session.lua:43: in function 'run'
        ...thub.com/neovim/.deps/usr/share/lua/5.1/nvim/session.lua:160: in function '_blocking_request'
        ...thub.com/neovim/.deps/usr/share/lua/5.1/nvim/session.lua:83: in function 'request'
        ./test/functional/helpers.lua:66: in function 'nvim'
        ./test/functional/helpers.lua:307: in function 'curbuf'
        ./test/functional/helpers.lua:323: in function 'curbuf_contents'
        ./test/functional/helpers.lua:343: in function 'expect'
        ...
        ...thub.com/neovim/.deps/usr/share/lua/5.1/busted/block.lua:150: in function 'execute'
        ...ithub.com/neovim/.deps/usr/share/lua/5.1/busted/init.lua:7: in function 'executor'
        ...ithub.com/neovim/.deps/usr/share/lua/5.1/busted/core.lua:276: in function <...ithub.com/neovim/.deps/usr/share/lua/5.1/busted/core.lua:276>
```
The commands themselves work fine when run manually, so it seems to be something specific to the testing.

mhinz replied:
Hmm, I can reproduce that.

jamessan replied:
`expand('%')` is the empty string when run in the lua test harness.  Something about that seems to break `expand('%:S')`.

jamessan replied:
Yeah, neovim segfaults. :)

mhinz commented:
Could you change the commit to this format?

```
vim-patch:7.4.1641

Problem:    Using unterminated string.
Solution:   Add NUL before calling vim_strsave_shellescape(). (James McCoy)

https://github.com/vim/vim/commit/5ca84ce4aa2832041f843e624c222bbc1f4d3e14
```

The short commit message of `vim-patch:7.4.1641` is important so it automatically gets picked up by https://neovim.io/doc/reports/vimpatch, the other stuff is more or less a convention. :)

mhinz commented on src/nvim/version.c:
```diff
@@ -69,6 +69,7 @@ static char *features[] = {
 
 // clang-format off
 static int included_patches[] = {
+  1641,
```
Usually all the commits not backported yet (so between 1641 and 1574) should be added as comment, to prevent weird merge conflicts later.. but looking at the next 4 lines people were obviously slacking recently.

mhinz commented:
Ehm...

```
:echo expand('%') | echo expand('%:r:S') | echo expand('%')
test/functional/legacy/105_filename_modifiers_spec.lua
'test/functional/legacy/105_filename_modifiers_spec'
test/functional/legacy/105_filename_modifiers_spec
```

jamessan commented:
Hmm, interesting.  It only applies for that command context.  The initial code I was working with resolves this, but I sent a simplified version to Bram.

I'll update this and send another patch to Bram.

jamessan committed:
vim-patch:7.4.1641

Problem:    Using unterminated string.
Solution:   Add NUL before calling vim_strsave_shellescape(). (James McCoy)

https://github.com/vim/vim/commit/5ca84ce4aa2832041f843e624c222bbc1f4d3e14

jamessan committed:
vim-patch:7.4.1643

Problem:    Terminating file name has side effects.
Solution:   Restore the character. (mostly by James McCoy, closes vim/vim#713)

https://github.com/vim/vim/commit/d4caf5c16a9f1c9477d426e58d8d3dc47ab5f066

jamessan committed:
vim-patch:7.4.1652

Problem:    Old style test for fnamemodify().
Solution:   Turn it into a new style test.

https://github.com/vim/vim/commit/610cc1b9b3c8104382f5506606c1f87118c28114

jamessan committed:
vim-patch:7.4.1654

Problem:    Crash when using expand('%:S') in a buffer without a name.
Solution:   Don't set a NUL. (James McCoy, closes vim/vim#714)

https://github.com/vim/vim/commit/52c6eaffd43a8c8865f8d6ed7cde0a8b137479e2

jamessan commented:
Updated with all the relevant patches from Vim.  One of the Travis builds failed in an unrelated test, but everything else is looking good.

```
[ RUN      ] ...ovim/neovim/test/functional/terminal/cursor_spec.lua @ 62: terminal cursor with number column is positioned correctly when focused

Warning: Screen changes have been received after the expected state was seen.

This is probably due to an indeterminism in the test. Try adding

`wait()` (or even a separate `screen:expect(...)`) at a point of possible

indeterminism, typically in between a `feed()` or `execute()` which is non-

synchronous, and a synchronous api call.

Note that sometimes a `wait` can trigger redraws and consequently generate more

indeterminism. If adding `wait` calls seems to increase the frequency of these

messages, try removing every `wait` call in the test.

If everything else fails, use Screen:redraw_debug to help investigate what is

  causing the problem.

      

stack traceback:

	./test/functional/ui/screen.lua:283: in function 'wait'

	./test/functional/ui/screen.lua:221: in function 'expect'

	...ovim/neovim/test/functional/terminal/cursor_spec.lua:64: in function <...ovim/neovim/test/functional/terminal/cursor_spec.lua:62>

./test/functional/ui/screen.lua:289: Row 2 didn't match.

Expected:

|  1 tty ready                                     |

|*  2 {1: }                                             |

|  3                                               |

|  4                                               |

|  5                                               |

|  6                                               |

|-- TERMINAL --                                    |

Actual:

|  1 tty ready                                     |

|*  2 rows: 6, cols: 46                             |

|  3 {1: }                                             |

|  4                                               |

|  5                                               |

|  6                                               |

|-- TERMINAL --                                    |

stack traceback:

	./test/functional/ui/screen.lua:289: in function 'wait'

	./test/functional/ui/screen.lua:221: in function 'expect'

	...ovim/neovim/test/functional/terminal/cursor_spec.lua:64: in function <...ovim/neovim/test/functional/terminal/cursor_spec.lua:62>
```

tarruda commented:
:+1: merged, thanks
